File: libsignal/service/src/main/java/org/whispersystems/signalservice/api/storage/StorageManifestKey.java
package org.whispersystems.signalservice.api.storage;

import java.util.Arrays;

/**
 * Key used to encrypt a manifest in the storage service.
 *
 * Created via {@link StorageKey#deriveManifestKey(long)}.
 */
public final class StorageManifestKey implements StorageCipherKey {

  private static final int LENGTH = 32;

  private final byte[] key;

  StorageManifestKey(byte[] key) {
    if (key.length != LENGTH) throw new AssertionError();

    this.key = key;
  }

  @Override
  public byte[] serialize() {
    return key.clone();
  }

  @Override
  public boolean equals(Object o) {
    if (o == null || o.getClass() != getClass()) return false;

    return Arrays.equals(((StorageManifestKey) o).key, key);
  }

  @Override
  public int hashCode() {
    return Arrays.hashCode(key);
  }
}

File: libsignal-service/src/main/java/org/whispersystems/signalservice/api/storage/StorageManifestKey.kt
package org.whispersystems.signalservice.api.storage

/**
 * Key used to encrypt a manifest in the storage service.
 *
 * Created via [StorageKey.deriveManifestKey].
 */
class StorageManifestKey(val key: ByteArray) : StorageCipherKey {
  init {
    check(key.size == 32)
  }

  override fun serialize(): ByteArray = key.clone()

  override fun equals(other: Any?): Boolean {
    if (this === other) return true
    if (javaClass != other?.javaClass) return false

    other as StorageManifestKey

    return key.contentEquals(other.key)
  }

  override fun hashCode(): Int {
    return key.contentHashCode()
  }
}
